src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs

```diff
@@ -0,0 +1,264 @@
+
+            string assemblyName = GetAssemblyName(metadataReader).FullName;
+
+            bool isAssemblyComVisible = IsComVisible(metadataReader, metadataReader.GetAssemblyDefinition());
+
+            foreach (TypeDefinitionHandle type in metadataReader.TypeDefinitions)
+            {
+                TypeDefinition definition = metadataReader.GetTypeDefinition(type);
+
+                if (TypeIsPublic(metadataReader, definition) && TypeIsClass(metadataReader, definition) && IsComVisible(metadataReader, definition, isAssemblyComVisible))
+                {
+                    string guid = GetTypeGuid(metadataReader, definition);
+
+                    if (clsidMap.ContainsKey(guid))
+                    {
+                        throw new BuildErrorException(Strings.ClsidMapConflictingGuids, clsidMap[guid].type, GetTypeName(metadataReader, definition), guid);
+                    }
+
+                    clsidMap.Add(guid,
+                        new ClsidEntry
+                        {
+                            type = GetTypeName(metadataReader, definition),
+                            assembly = assemblyName
+                        });
+                }
+            }
+
+            using (StreamWriter writer = File.CreateText(clsidMapPath))
+            {
+                writer.Write(JsonConvert.SerializeObject(clsidMap));
+            }
+        }
+
+        private static bool TypeIsClass(MetadataReader metadataReader, TypeDefinition definition)
+        {
+            if ((definition.Attributes & TypeAttributes.Interface) != 0)
+            {
+                return false;
+            }
+
+            EntityHandle baseTypeEntity = definition.BaseType;
+            if (baseTypeEntity.Kind == HandleKind.TypeReference)
+            {
+                TypeReference baseClass = metadataReader.GetTypeReference((TypeReferenceHandle)baseTypeEntity);
+                if (baseClass.ResolutionScope.Kind == HandleKind.AssemblyReference && ReferenceToCoreLibrary(metadataReader, (AssemblyReferenceHandle)baseClass.ResolutionScope))
+                {
+                    if (HasTypeName(metadataReader, baseClass, "System", "ValueType") || HasTypeName(metadataReader, baseClass, "System", "Enum"))
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        private static readonly string[] s_CoreLibraryNames = { "mscorlib", "System.Runtime", "netstandard" };
+
+        private static bool ReferenceToCoreLibrary(MetadataReader reader, AssemblyReferenceHandle assembly)
+        {
+            return s_CoreLibraryNames.Contains(reader.GetString(reader.GetAssemblyReference(assembly).Name));
+        }
+
+        private static bool TypeIsPublic(MetadataReader reader, TypeDefinition type)
+        {
+            if ((type.Attributes & TypeAttributes.VisibilityMask) != TypeAttributes.Public)
```

Comment:
This should be `==`, not `!=`.

Consider switching over `type.Attributes & typeAttributes.VisibilityMask`